### PR TITLE
SHA3 xof shake128 and shake256 

### PR DESCRIPTION
**Pull request by immoschuett**
Hi @staticfloat,

i added some implementations of the extended output function shake, since i needed them for my work.  I thought this module could take advantage of it.

This is my first pull request, so i'm sorry if anything is not along the usual workflow. Plealse give me feedback!

kind regards.

### CODE REVIEW

File: src/SHA.jl [hunk 1]
```diff
@@ -83,7 +87,7 @@ for (f, ctx) in [(:sha1, :SHA1_CTX),
         See also [`$($ctx)`](@ref).
         """
         function $f(data::AbstractBytes)
-            ctx = $ctx()
+            _ = $ctx()
```
Comment (staticfloat) [inline src/SHA.jl:90]: I don't think this change is correct

File: src/SHA.jl [hunk 2]
```diff
@@ -112,7 +116,7 @@ for (f, ctx) in [(:sha1, :SHA1_CTX),
         Hash data from io using `$($f)` algorithm.
         """
         function $f(io::IO, chunk_size=4*1024)
-            ctx = $ctx()
+            _ = $ctx()
```
Comment (staticfloat) [inline src/SHA.jl:119]: Same with this change, I don't think it's correct

**Review by staticfloat — CHANGES_REQUESTED** [review]
Thanks for your contribution! 

These changes seem to have broken the tests, so first resolve that.  Next, you'll need to implement a number of tests to ensure that the implementation is correct.  Then I can review again.

**Comment by immoschuett** [comment]
Thanks for the fast feedback, i fixed the issues and will add some tests regarding shake soon.

**Comment by immoschuett** [comment]
Hi @staticfloat

I added some basic testing. 
Yet, I did not add any documentation, only comments, but I can work on that later on. I wanted to wait for your first review, because i expect, that some Structs from SHAKE can/should be moved somwhere else.

File: src/shake.jl [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        for i in 1:5
+            bc[i] = state[i] ⊻ state[i + 5] ⊻ state[i + 10] ⊻ state[i + 15] ⊻ state[i + 20]
+        end
+        for i in 0:4
+            temp = bc[rem(i + 4, 5) + 1] ⊻ L64(1, bc[rem(i + 1, 5) + 1])
+            j = 0
+            while j <= 20
+                state[Int(i + j + 1)] = state[i + j + 1] ⊻ temp
+                j += 5
+            end
```
Comment (staticfloat) [inline src/shake.jl:51]: Perhaps this should be a `for` loop, just to make this kind of stuff easier for the compiler to vectorize?

```suggestion
            for j in 0:5:20
                state[Int(i + j + 1)] = state[i + j + 1] ⊻ temp
            end
```

File: src/shake.jl [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        j = 0
+        while j <= 20
+            for i in 1:5
+                bc[i] = state[i + j]
+            end
+            for i in 0:4
+                state[j + i + 1] = state[j + i + 1] ⊻ (~bc[rem(i + 1, 5) + 1] & bc[rem(i + 2, 5) + 1])
+            end
+            j += 5
+        end
```
Comment (staticfloat) [inline src/shake.jl:71]: Same here

```suggestion
        for j in 0:5:20
            for i in 1:5
                bc[i] = state[i + j]
            end
            for i in 0:4
                state[j + i + 1] = state[j + i + 1] ⊻ (~bc[rem(i + 1, 5) + 1] & bc[rem(i + 2, 5) + 1])
            end
        end
```

File: test/runtests.jl [hunk 5]
```diff
@@ -141,3 +141,22 @@ end
         @test_throws MethodError f(UInt32[0x23467, 0x324775])
     end
 end
+
+@testset "SHAKE" begin
+    @testset "shake128" begin
+        for  (k,v) in SHA128test
```
Comment (staticfloat) [inline test/runtests.jl:147]:
```suggestion
        for (k,v) in SHA128test
```

File: test/runtests.jl [hunk 6]
```diff
@@ -141,3 +141,22 @@ end
...
+            @test SHA.shake128(hex2bytes(k[1]),k[2]) == hex2bytes(v)
+        end
+        @test SHA.shake128(b"",UInt(16)) == hex2bytes("7f9c2ba4e88f827d616045507605853e")
+    end
+
+    @testset "shake256" begin
+        for  (k,v) in SHA256test
```
Comment (staticfloat) [inline test/runtests.jl:154]:
```suggestion
        for (k,v) in SHA256test
```

**Review by staticfloat — APPROVED** [review]

File: src/shake.jl (hunk 3 above)
Comment (immoschuett) [reply src/shake.jl]: yes, i implemented this all together with the docstrings in my last commit.

File: src/shake.jl [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        context.used = true
+        p+=blocklen(T)
+        digest!(context,d-blocklen(T),p)
+        return 
+    end
+end
+"""
+            shake128(data::AbstractBytes,d::UInt)
+
+        Hash data using the `shake128` algorithm and return the first d resulting bytes.
```
Comment (staticfloat) [inline src/shake.jl:112]: You don't need to indent these so heavily:

```suggestion
    shake128(data::AbstractBytes,d::UInt)

Hash data using the `shake128` algorithm and return the first d resulting bytes.
```

**Comment by staticfloat** [comment]
LGTM!  Thanks!

**Pull request merged**